### DISCUSSION TITLE
How does the interface filter (the JS based one built in to the table) work?

### DISCUSSION THREAD

**Discussion by cpmills1975** [netbox-community/netbox#5647] [General]
Hi All

I've been trying to understand how to use the (JavaScript based) filter that is built in to the interface table in 2.10.3 (and perhaps earlier - I've skipped many versions!). I understand it accepts a RegEx.

Take for example a Juniper Switch with ge-0/0/[0-47].
* If I type 'g' in this field, I get nothing.
* If I type 'g*' I get everything.
* If I type 'ge' in this field, I get nothing
* If I type 'ge*' in this field, I still get nothing!
* I want to find all interfaces that end in '.0' (JunOS 'unit 0' interfaces) - I want to delete them - I've tried searching for '*.0', '.0$', all sorts, I can never get it to return anything sensible.  Hmm - maybe I didn't try 'backslash dot 0', should that escape the dot?
* If I have the JavaScript console open, it is extremely easy to get it to throw many many errors!

Some practical examples would be helpful - I know at some point, the users I support are going to ask...

Many thanks

**Comment by cpmills1975** [comment]
Hmm. I might actually be seeing a browser issue as I am now unable to replicate my problems at home...

**Comment by cpmills1975** [comment]
I've gotten to the bottom of this and will document here for others in case it's useful.

I'm using the netbox-docker image. At startup, it calls manage.py collectstatic to copy the static files to a dedicated static volume. When the volume was created, it was so with NetBox 2.10.0 as that was the latest image I had lying around and was after the release of 2.10.2 in which 442dee72b59f727109ab993db5f17b64f61e649f changed the way the filter worked.

When I upgraded to 2.10.3, the file modification date on static/js/interface_filtering.js was older than the date collectstatic last copied the files to the volume and hence it wasn't updated. The code was expecting to find an id in the form interface_ge/0/0/0, yet the template was generating the data-name field as per the changes made to dcim/tables/devices.py in 442dee72b59f727109ab993db5f17b64f61e649f hence it did nothing.

I fixed this by deleting the entire static directory and re-running manage.py collectstatic to re-copy the files. This also had a side effect in that it removed the font-awesome stuff that is no longer used but hadn't been deleted. It feels like this is a potential bug in the netbox-docker image and I'll raise a bug on https://github.com/netbox-community/netbox-docker however, is there any reason at all to retain any of the contents of the static volume between upgrades?  I'm wondering whether it should just be blatted as a matter of course.

**Reply by candlerb** [reply]
Ouch. Thanks for the detailed analysis.

I fully agree with you: I think that the docker container ought to `collectstatic` to its own local storage (not a persistent volume), so that every time you replace the version with a newer one, it gets completely fresh set of static files.  Even better would be for `collectstatic` to be run at the time that the docker image is built, not at [docker run time](https://github.com/netbox-community/netbox-docker/blob/release/docker/docker-entrypoint.sh#L64).

I think that a persistent volume should only be required for media, scripts and reports (and of course postgres).

Issue appears to be [here](https://github.com/netbox-community/netbox-docker/blob/release/docker-compose.yml#L19) - presumably there was a reason why `/opt/netbox/netbox/static` was given a persistent volume, but I can't think what it would be.

**Comment by hSaria** [comment]
Hi @cpmills1975 

I don't know if you still need help on this, but hopefully the following will shed some light.

The interface filter uses regular expressions, not shell expansion. You may be used to shell expansion in bash (i.e. `ls -l *.log`). Regular expressions are used in `grep` if you're familiar with that.

Example regular expressions:
* `.e-0/0/0`: The `.` means "any single character", so this will match `ge-0/0/0` as well as `xe-0/0/0`, and even `ze-0/0/0` (whatever that is).
* `.*/0/0`: `*` is a quantifier of "zero or more of what came before", so `.*` will mean "zero or more of any single character". Another common quantifier is `+` which is "one or more of what came before".
* `ge-[0-9]/0/0`: Here, a set is used match a range of values (0 to 9), so you can match `ge-0/0/0` as well as `ge-1/0/0`.
* `ge-[0-14-59]/0/0`: In this set, you have two ranges of values (0 to 1 and 4 to 5) as well as a value (9). I'm just showing you here that the set can include multiple values/ranges. Another way to write the same set is `ge-[01459]/0/0`.
* `ge-[0-9]+/0/0`: This will try to match one or more digits in the FPC (so it would match 1 as well as 11).
*  `(ge|xe)-0/0/0`: If you want to express an OR condition (i.e. this value or that value), you can group the values in parentheses and use `|`. This example would match `ge-0/0/0` and `xe-0/0/0`. You can create more than one OR, like `(ge|xe|fxp0)`.
* `/0$`: If you want to match all ports with number 0, you might be tempted to just type `/0`, but that will match `ge-0/0/1` (PIC is matching in the middle). To get around that, you can use `$` to indicate the end of the string. In this example, an interface must end with `/0` to match. Similarly, you can use `^` to indicate the start of the string, like `^ge`.

Note that this is **not** exclusive to NetBox; this is just how regular expressions work. There are a lot of resources online with more complex examples, but I think these would suffice.
